target/arc/fpu-helper.c

```diff
@@ -0,0 +1,938 @@
+/*
+ * QEMU ARC CPU
+ *
+ * Copyright (c) 2021 Synppsys Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see
+ * http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+
+#include "qemu/osdep.h"
+#include "cpu.h"
+#include "exec/exec-all.h"
+#include "exec/helper-proto.h"
+#include "fpu/softfloat.h"
+#include "fpu.h"
+#include "qemu/log.h"
+
+#define FWE	31
+#define CIX 	12
+#define CUF 	11
+#define COF 	10
+#define CDZ 	 9
+#define CIV 	 8
+#define CAL 	 7
+#define IX  	 4
+#define UF  	 3
+#define OF  	 2
+#define DZ  	 1
+#define IV  	 0
+
+#define BIT_READ(LOC, BIT)  ((LOC >> BIT) & 1)
+#define BIT_WRITE(LOC, BIT, VAL)  LOC = (LOC & (~(1 << BIT))) | ((VAL & 1) << BIT)
+
+#define READ_EXCP_FLAG(LOC, BIT) ((LOC & BIT) != 0)
+#define WRITE_EXCP_FLAG(LOC, BIT, VALUE) LOC = ((LOC & (~BIT)) | ((VALUE & 1) * BIT))
+
+static uint8_t fpr_width;
+uint8_t vfp_width;
+/* TODO: Maybe is not needed */
+static uint8_t fpr_per_vector_operand;
+static uint8_t vfp_max_length;
+
+void
+init_fpu(CPUARCState *env, bool fp_dp, bool fp_wide, bool fp_hp)
+{
+  uint8_t index1 = ((fp_dp != 0) * 2) + (fp_wide != 0);
+  const uint8_t fpr_widths [] = { 32, 32, 64, 64 };
+  const uint8_t vfp_widths [] = { 32, 64, 64, 128 };
+
+  fpr_width = fpr_widths[index1];
+  vfp_width = vfp_widths[index1];
+
+  fpr_per_vector_operand = (vfp_width / fpr_width);
+  vfp_max_length = fpr_per_vector_operand * (vfp_width / 32) * (fp_hp != 0 ? 2 : 1);
+
+  uint32_t tmp = 0;
+  set_default_nan_mode(0, &env->fp_status);
+  set_float_rounding_mode(float_round_nearest_even, &env->fp_status);
```

Comment:
Please add a comment where you refer to PRM's default rounding mode for FP section.